doc/developer/design/20230615_webhook_source.md

```diff
@@ -0,0 +1,232 @@
+latency of the events works very well with the premise of being an "operational" data warehouse.
+
+> **_Note_** To achieve the goal of "make it easier to get data in Materialize" we also considered
+building support for Fivetran. We may still do this in the future, but the amount of work to get
+webhooks working is significantly less, and the frequency of data is more suited to the goals of
+Materialize.
+
+## Goals
+
+1. Provide URLs that users can push data to. In Materialize this data is represented as a source
+with two columns, `headers` and `body`.
+2. Support an extensible syntax similar to `CREATE SOURCE <name> FROM HTTP REQUESTS` that allows
+users to generate the aforementioned URLs.
+3. Handle at least a total of 100 requests per-second across all `HTTP REQUESTS` sources.
+5. Code complete by August 15th, 2023.
+
+
+## Non-Goals
+
+The following are all features we want to support, but are out of scope in the immediate term.
+
+* Extracting keys from the data for `UPSERT`-like behavior.
+* Validation that the request comes from a trusted source.
+* Custom filtering of headers.
+* Manipulating the incoming data, splitting a single request into any kind of "sub-sources".
+
+## Overview
+
+Our implementation will mostly exist in the adapter, and we can re-use existing APIs on the
+`storage-client` to handle the persistence of data.
+
+We'll update the SQL parser to support our new kind of `HTTP REQUESTS` source, and add
+a new endpoint to `environmentd`'s HTTP server `/api/webhook/:id`. When a user creates this new
+kind of source, we'll generate an id, persist the new `Source` in the catalog, use the
+`storage-client` to create a new collection, and finally return a well formatted URL to the user.
+
+When we receive a event we'll use the created `SessionClient` to send the raw `headers` and `body`
+to the coordinator, pack the data into a row, and append it to the necessary collection using a
+new monotonic append API on the `StorageController`.
+
+## Detailed description
+
+We'll update the parser to support creating a new kind of Source, `HTTP REQUESTS`, for which you can
+specify a `FORMAT`, which denotes how we decode the body of a request, the only supported format
+for the time being is `BYTES`.
+
+```
+CREATE SOURCE <name> FROM HTTP REQUESTS FORMAT BYTES;
+```
+
+After executing this request, we'll create a source with the following columns:
+
+| name    | type              |
+|---------|-------------------|
+| headers | map[text -> text] |
+| body    | bytea             |
+
+Using `bytea` might not be optimal from a storage perspective (e.g. no chance for MFP Pushdown) but
+it makes this initial attempt at webhooks maximally compatible. On the completion of the statement
+we'll return an ID to the user, which they would specify in the webhook URL. The returned ID would
+be 6 character base62 encoded string, a pattern commonly used in URL shorteners.
```

Comment:
Wait, why not just use the source's global ID as the ID?

Reply:
I left a comment about this above, but basically I think using an external ID gives us more flexibility. The user would be able to drop and recreate the source, even rename it, without having to update the applications that send events to it, assuming we can re-use IDs. I think this maps pretty well to our existing sources, where users can drop and rename them without having to modify the services they pull data from.

Reply:
We can't re-use IDs. Changing that would require some thought.